<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>GEENOME — GEENOME • GEENOME</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="GEENOME — GEENOME"><meta property="og:description" content="TGEENOME is a 16S rRNA metagenomic analysis tool that encompasses nearly all steps of downstream analysis. These steps include preprocessing to filter zero-inflated and low abundance data, 
exploring the dataset through various plots, calculating alpha and beta diversity, generating ordination plots, testing null hypotheses, 
and applying a novel differential expression method using generalized estimating equations. 
The results of all the aforementioned steps are compiled into a single, detailed PDF file within a well-organized folder structure."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GEENOME</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>GEENOME</h1>
    
    <div class="hidden name"><code>GEENOME.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>TGEENOME is a 16S rRNA metagenomic analysis tool that encompasses nearly all steps of downstream analysis. These steps include preprocessing to filter zero-inflated and low abundance data, 
exploring the dataset through various plots, calculating alpha and beta diversity, generating ordination plots, testing null hypotheses, 
and applying a novel differential expression method using generalized estimating equations. 
The results of all the aforementioned steps are compiled into a single, detailed PDF file within a well-organized folder structure.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">GEENOME</span><span class="op">(</span></span>
<span>  <span class="va">physeq</span>,</span>
<span>  <span class="va">variables</span>,</span>
<span>  <span class="va">id</span>,</span>
<span>  <span class="va">sample_var</span>,</span>
<span>  <span class="va">group_var</span>,</span>
<span>  <span class="va">out_cut</span>,</span>
<span>  <span class="va">zero_cut</span>,</span>
<span>  <span class="va">lib_cut</span>,</span>
<span>  <span class="va">neg_lb</span>,</span>
<span>  <span class="va">model_form</span>,</span>
<span>  <span class="va">alpha</span>,</span>
<span>  <span class="va">n_cl</span>,</span>
<span>  <span class="va">prv_cut</span>,</span>
<span>  <span class="va">AlphaBetaDiversity</span>,</span>
<span>  <span class="va">color_label</span>,</span>
<span>  <span class="va">BetaDiversity.distance</span>,</span>
<span>  <span class="va">shape_label</span>,</span>
<span>  <span class="va">axes</span>,</span>
<span>  <span class="va">permanova.distance</span>,</span>
<span>  <span class="va">permanova.strata</span>,</span>
<span>  <span class="va">permanova.permutation_number</span>,</span>
<span>  <span class="va">adj_pvalue</span>,</span>
<span>  <span class="va">GEE_analysis</span>,</span>
<span>  <span class="va">fill_Alpha</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>physeq</dt>
<dd><p>Phyloseq format. The data input in phyloseq format.</p></dd>


<dt>variables</dt>
<dd><p>Character. The co-founders exist in metadata that affect on the taxa. in variables &amp; cofounders for RCM</p></dd>


<dt>id</dt>
<dd><p>Character. Individual id (personal id for each participant even if repeated measures in time series)</p></dd>


<dt>sample_var</dt>
<dd><p>Character. The name of column storing sample IDs.</p></dd>


<dt>group_var</dt>
<dd><p>Character. The name of the group indicator.</p></dd>


<dt>out_cut</dt>
<dd><p>Numerical. fraction between 0 and 1.</p></dd>


<dt>zero_cut</dt>
<dd><p>Numerical. fraction between 0 and 1.</p></dd>


<dt>lib_cut</dt>
<dd><p>Numeric. Samples with library size less than lib_cut are not included in the analysis.</p></dd>


<dt>neg_lb</dt>
<dd><p>Logical. TRUE indicates a taxon would be classified as a structural zero.</p></dd>


<dt>model_form</dt>
<dd><p>Character. Just detect the variables you need to enter in the formula.</p></dd>


<dt>alpha</dt>
<dd><p>Numeric. alpha cutoff</p></dd>


<dt>n_cl</dt>
<dd><p>Numeric.</p></dd>


<dt>prv_cut</dt>
<dd><p>Numeric.</p></dd>


<dt>AlphaBetaDiversity</dt>
<dd><p>Logical. If you need to apply alpha and beta diversity steps or not. (the phyloseq must have tax_table)</p></dd>


<dt>color_label</dt>
<dd><p>Character. the variable or co-founder that used in coloring the plots. in the interested variable &amp; in RCM</p></dd>


<dt>BetaDiversity.distance</dt>
<dd><p>Character. choose from "distanceMethodList" function</p></dd>


<dt>shape_label</dt>
<dd><p>Character. the variable or co-founder that used in shape of the plots. in RCM only</p></dd>


<dt>axes</dt>
<dd><p>Numeric. You need to detect the axes before plotting the ordination after scree plot</p></dd>


<dt>permanova.distance</dt>
<dd><p>Character. choose from "distanceMethodList" function</p></dd>


<dt>permanova.strata</dt>
<dd><p>Character. add strata if you need. The default is NULL</p></dd>


<dt>permanova.permutation_number</dt>
<dd><p>Numeric. The number of permutation.</p></dd>


<dt>adj_pvalue</dt>
<dd><p>Character. The choosen method in adjust p-value. choose from c("holm", "hochberg", "hommel", "bonferroni", "BH", "BY", "fdr", "none")</p></dd>


<dt>GEE_analysis</dt>
<dd><p>Logical. Apply GEE downstream optionality if you don't need to re-analysis your data.</p></dd>


<dt>fill_Alpha</dt>
<dd><p>Character. choose your interested taxonomic level (Please make sure tax_table() in physeq has that's level)</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A PDF file in the same directory called "GEENOME_plots.pdf"</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>GEENOME</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://dx.plos.org/10.1371/journal.pone.0061217" class="external-link">phyloseq</a></span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># load A two-week diet swap study between western (USA) and traditional (rural Africa) diets (Lahti et al. 2014).</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">dietswap</span>, package <span class="op">=</span> <span class="st">"microbiome"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">phyloseq_data</span> <span class="op">&lt;-</span> <span class="va">dietswap</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># call library GEENOME</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">GEENOME</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># detect various parameters</span></span></span>
<span class="r-in"><span><span class="co"># enter your phyloseq_data variable to be named "physeq"</span></span></span>
<span class="r-in"><span><span class="va">physeq</span> <span class="op">=</span> <span class="va">phyloseq_data</span> </span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># make sure from the column names in sample_data(physeq)</span></span></span>
<span class="r-in"><span><span class="va">variables</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"bmi_group"</span>,<span class="st">"timepoint.within.group"</span><span class="op">)</span> <span class="co"># variables &amp; cofounders for RCM</span></span></span>
<span class="r-in"><span><span class="va">color_label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"timepoint"</span><span class="op">)</span> <span class="co"># is the interested variable &amp; in RCM</span></span></span>
<span class="r-in"><span><span class="va">shape_label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sex"</span><span class="op">)</span> <span class="co"># in RCM only</span></span></span>
<span class="r-in"><span><span class="va">id</span> <span class="op">=</span> <span class="st">"subject"</span> <span class="co"># individual id (personal id for each participant even if repeated measures in time series)</span></span></span>
<span class="r-in"><span><span class="va">sample_var</span> <span class="op">=</span> <span class="st">"sample"</span> <span class="co"># sample ID (equal to rownames of physeq@sam_data)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Apply GEE downstream optionality if you don't need to re-analysis your data. </span></span></span>
<span class="r-in"><span><span class="va">GEE_analysis</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span> </span></span>
<span class="r-in"><span><span class="va">model_form</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"bmi_group"</span>,<span class="st">"timepoint.within.group"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># your need to make sure that phyloseq contain tax_table() before applying alpha &amp; beta diversity</span></span></span>
<span class="r-in"><span><span class="va">AlphaBetaDiversity</span> <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span><span class="va">BetaDiversity.distance</span> <span class="op">&lt;-</span> <span class="st">"bray"</span> <span class="co"># choose from these distances =&gt; distanceMethodList https://joey711.github.io/phyloseq/distance.html</span></span></span>
<span class="r-in"><span><span class="va">fill_Alpha</span> <span class="op">=</span> <span class="st">"Family"</span> <span class="co"># choose your interested taxonomic level (Please make sure tax_table() in physeq has that's level) </span></span></span>
<span class="r-in"><span><span class="va">axes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span> <span class="co"># You need to detect the axes before plotting the ordination after scree plot</span></span></span>
<span class="r-in"><span><span class="va">permanova.distance</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"bray"</span><span class="op">)</span> <span class="co"># https://rdrr.io/bioc/phyloseq/man/distanceMethodList.html</span></span></span>
<span class="r-in"><span><span class="va">Permanova.strata</span> <span class="op">=</span> <span class="cn">NULL</span> <span class="co"># add strata if you need</span></span></span>
<span class="r-in"><span><span class="va">permanova.permutation_number</span> <span class="op">&lt;-</span> <span class="fl">9999</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Preprocessing parameters</span></span></span>
<span class="r-in"><span><span class="va">group_var</span> <span class="op">=</span> <span class="cn">NULL</span></span></span>
<span class="r-in"><span><span class="va">out_cut</span> <span class="op">=</span> <span class="fl">0.05</span></span></span>
<span class="r-in"><span><span class="va">zero_cut</span> <span class="op">=</span> <span class="fl">0.9</span></span></span>
<span class="r-in"><span><span class="va">lib_cut</span> <span class="op">=</span> <span class="fl">1000</span></span></span>
<span class="r-in"><span><span class="va">neg_lb</span> <span class="op">=</span> <span class="cn">FALSE</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">adj_pvalue</span> <span class="op">&lt;-</span> <span class="st">"BH"</span> <span class="co"># choose from these =&gt; c("holm", "hochberg", "hommel", "bonferroni", "BH", "BY", "fdr", "none")</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Run GEENOME</span></span></span>
<span class="r-in"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">GEENOME</span><span class="op">(</span><span class="va">physeq</span>, <span class="va">variables</span>, <span class="va">id</span>, <span class="va">sample_var</span>, <span class="va">group_var</span>, <span class="va">out_cut</span>, <span class="va">zero_cut</span>, </span></span>
<span class="r-in"><span>              <span class="va">lib_cut</span>, <span class="va">neg_lb</span>, <span class="va">model_form</span>, <span class="va">alpha</span>, <span class="va">n_cl</span>, <span class="va">prv_cut</span>,<span class="va">AlphaBetaDiversity</span>,</span></span>
<span class="r-in"><span>              <span class="va">color_label</span>,<span class="va">BetaDiversity.distance</span>,<span class="va">shape_label</span>,<span class="va">axes</span>,<span class="va">permanova.distance</span>,</span></span>
<span class="r-in"><span>              <span class="va">permanova.strata</span>,<span class="va">permanova.permutation_number</span>,<span class="va">adj_pvalue</span>,<span class="va">GEE_analysis</span>,<span class="va">fill_Alpha</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "dplyr"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] ‘1.0.10’</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "nlme"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] ‘3.1.159’</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "ggplot2"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] ‘3.4.3’</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "compositions"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] ‘2.0.4’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Welcome to compositions, a package for compositional data analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Find an intro with "? compositions"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "plyr"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] ‘1.8.8’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ------------------------------------------------------------------------------</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> You have loaded plyr after dplyr - this is likely to cause problems.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> If you need functions from both plyr and dplyr, please load plyr first, then dplyr:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> library(plyr); library(dplyr)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ------------------------------------------------------------------------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "tidyverse"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] ‘1.3.2’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ── <span style="font-weight: bold;">Attaching packages</span> ─────────────────────────────────────── tidyverse 1.3.2 ──</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">tibble </span> 3.1.8     <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">purrr  </span> 0.3.5</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">tidyr  </span> 1.2.1     <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">stringr</span> 1.4.1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">readr  </span> 2.1.3     <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">forcats</span> 0.5.2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ── <span style="font-weight: bold;">Conflicts</span> ────────────────────────────────────────── tidyverse_conflicts() ──</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">plyr</span>::<span style="color: #00BB00;">arrange()</span>   masks <span style="color: #0000BB;">dplyr</span>::arrange()</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">nlme</span>::<span style="color: #00BB00;">collapse()</span>  masks <span style="color: #0000BB;">dplyr</span>::collapse()</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">purrr</span>::<span style="color: #00BB00;">compact()</span>  masks <span style="color: #0000BB;">plyr</span>::compact()</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">plyr</span>::<span style="color: #00BB00;">count()</span>     masks <span style="color: #0000BB;">dplyr</span>::count()</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">plyr</span>::<span style="color: #00BB00;">failwith()</span>  masks <span style="color: #0000BB;">dplyr</span>::failwith()</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">dplyr</span>::<span style="color: #00BB00;">filter()</span>   masks <span style="color: #0000BB;">stats</span>::filter()</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">plyr</span>::<span style="color: #00BB00;">id()</span>        masks <span style="color: #0000BB;">dplyr</span>::id()</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">dplyr</span>::<span style="color: #00BB00;">lag()</span>      masks <span style="color: #0000BB;">stats</span>::lag()</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">plyr</span>::<span style="color: #00BB00;">mutate()</span>    masks <span style="color: #0000BB;">dplyr</span>::mutate()</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">plyr</span>::<span style="color: #00BB00;">rename()</span>    masks <span style="color: #0000BB;">dplyr</span>::rename()</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">plyr</span>::<span style="color: #00BB00;">summarise()</span> masks <span style="color: #0000BB;">dplyr</span>::summarise()</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">plyr</span>::<span style="color: #00BB00;">summarize()</span> masks <span style="color: #0000BB;">dplyr</span>::summarize()</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "gsubfn"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] ‘0.7’</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "zCompositions"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] ‘1.4.0.1’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘MASS’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:dplyr’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     select</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘NADA’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:compositions’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     cor</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:stats’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     cor</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "compositions"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] ‘2.0.4’</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "grid"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] ‘4.2.1’</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "gridExtra"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] ‘2.3’</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "nlme"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] ‘3.1.159’</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "optiscale"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] ‘1.2.2’</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "propr"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] ‘4.2.6’</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "webshot"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] ‘0.5.4’</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "ftExtra"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] ‘0.4.0’</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "flextable"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] ‘0.8.3’</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "caret"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] ‘6.0.93’</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "stringr"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] ‘1.4.1’</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "DT"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] ‘0.26’</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "htmlwidgets"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] ‘1.5.4’</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "geepack"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] ‘1.3.9’</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "ggpubr"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] ‘0.5.0’</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "vegan"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] ‘2.6.4’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> This is vegan 2.6-4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "scales"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] ‘1.2.1’</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "phyloseq"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] ‘1.40.0’</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "RCM"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] ‘1.12.0’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Registered S3 method overwritten by 'quantmod':</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   method            from</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   as.zoo.data.frame zoo </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "data.table"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] ‘1.14.4’</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "microbiome"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] ‘1.18.0’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> microbiome R package (microbiome.github.com)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  Copyright (C) 2011-2022 Leo Lahti, </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     Sudarshan Shetty et al. &lt;microbiome.github.io&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "heatmaply"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] ‘1.4.2’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘plotly’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:flextable’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     highlight, style</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:MASS’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     select</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:plyr’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     arrange, mutate, rename, summarise</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:ggplot2’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     last_plot</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:stats’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     filter</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:graphics’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     layout</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘viridis’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:scales’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     viridis_pal</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Registered S3 method overwritten by 'dendextend':</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   method     from </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   rev.hclust vegan</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Registered S3 method overwritten by 'seriation':</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   method         from </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   reorder.hclust vegan</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ======================</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Welcome to heatmaply version 1.4.2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Type citation('heatmaply') for how to cite the package.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Type ?heatmaply for the main documentation.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The github page is: https://github.com/talgalili/heatmaply/</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Please submit your suggestions and bug-reports at: https://github.com/talgalili/heatmaply/issues</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> You may ask questions at stackoverflow, use the r and heatmaply tags: </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 	 https://stackoverflow.com/questions/tagged/heatmaply</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ======================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "permute"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] ‘0.9.7’</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> !!!&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt; your output will be in this directory: /home/hp/Documents/GitHub/GEENOME/docs/reference </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> !!!&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt; make sure that this script and your data in the same directory </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> !!!&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt; The output plots in details will be saved in separated folder called *GEENOME_plots*Done &gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt; 1.Done Preprocessing </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Done &gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt; 2.Done Explore the data </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> `set.seed(1)` was used to initialize repeatable random subsampling.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Please record this for your records so others can reproduce.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Try `set.seed(1); .Random.seed` for the full vector</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 1 samples removedbecause they contained fewer reads than `sample.size`.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Up to first five removed samples are: </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Sample-56</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ...</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>cannot compute exact p-value with ties</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>cannot compute exact p-value with ties</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>cannot compute exact p-value with ties</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>cannot compute exact p-value with ties</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>cannot compute exact p-value with ties</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>cannot compute exact p-value with ties</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>cannot compute exact p-value with ties</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>cannot compute exact p-value with ties</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>cannot compute exact p-value with ties</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>cannot compute exact p-value with ties</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>cannot compute exact p-value with ties</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>cannot compute exact p-value with ties</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>cannot compute exact p-value with ties</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>cannot compute exact p-value with ties</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>cannot compute exact p-value with ties</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Done &gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt; 3.Done Alpha diversity </span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>The melt generic in data.table has been passed a list and will attempt to redirect to the relevant reshape2 method; please note that reshape2 is deprecated, and this redirection is now deprecated as well. To continue using melt methods from reshape2 while both libraries are attached, e.g. melt.list, you can prepend the namespace like reshape2::melt(sub_dist). In the next version, this warning will become an error.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>The melt generic in data.table has been passed a matrix and will attempt to redirect to the relevant reshape2 method; please note that reshape2 is deprecated, and this redirection is now deprecated as well. To continue using melt methods from reshape2 while both libraries are attached, e.g. melt.list, you can prepend the namespace like reshape2::melt(physeq_Distance). In the next version, this warning will become an error.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Done &gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt; 4.Done Beta diversity </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Square root transformation</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Wisconsin double standardization</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 0 stress 0.2133623 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 1 stress 0.2178022 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 2 stress 0.2211338 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 3 stress 0.2209762 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 4 stress 0.416746 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 5 stress 0.2136226 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ... Procrustes: rmse 0.01152948  max resid 0.1018301 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 6 stress 0.2245528 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 7 stress 0.2149561 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 8 stress 0.2136543 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ... Procrustes: rmse 0.01401871  max resid 0.1573673 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 9 stress 0.2175167 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 10 stress 0.2255632 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 11 stress 0.2187455 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 12 stress 0.2193386 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 13 stress 0.2274358 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 14 stress 0.2163268 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 15 stress 0.2264926 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 16 stress 0.2261255 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 17 stress 0.2267637 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 18 stress 0.2204196 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 19 stress 0.2203056 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 20 stress 0.2196194 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> *** Best solution was not repeated -- monoMDS stopping criteria:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     18: stress ratio &gt; sratmax</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      2: scale factor of the gradient &lt; sfgrmin</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Square root transformation</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Wisconsin double standardization</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 0 stress 0.2133623 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 1 stress 0.2219124 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 2 stress 0.213183 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ... New best solution</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ... Procrustes: rmse 0.01938682  max resid 0.1554535 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 3 stress 0.2294731 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 4 stress 0.2271144 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 5 stress 0.2182722 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 6 stress 0.2233627 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 7 stress 0.2133671 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ... Procrustes: rmse 0.01646453  max resid 0.1245292 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 8 stress 0.2127607 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ... New best solution</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ... Procrustes: rmse 0.0174943  max resid 0.1676176 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 9 stress 0.221265 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 10 stress 0.2212965 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 11 stress 0.2183751 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 12 stress 0.2222367 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 13 stress 0.2250929 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 14 stress 0.2266284 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 15 stress 0.2202835 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 16 stress 0.2229953 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 17 stress 0.2237853 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 18 stress 0.2163237 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 19 stress 0.2174355 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 20 stress 0.2211585 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> *** Best solution was not repeated -- monoMDS stopping criteria:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      2: no. of iterations &gt;= maxit</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     16: stress ratio &gt; sratmax</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      2: scale factor of the gradient &lt; sfgrmin</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> No eigenvalues found in ordination</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NULL</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>The following aesthetics were dropped during statistical transformation: <span style="color: #00BB00;">shape</span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> This can happen when ggplot fails to infer the correct grouping structure in</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   the data.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Did you forget to specify a `group` aesthetic or to convert a numerical</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   variable into a factor?</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>The following aesthetics were dropped during statistical transformation: <span style="color: #00BB00;">shape</span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> This can happen when ggplot fails to infer the correct grouping structure in</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   the data.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Did you forget to specify a `group` aesthetic or to convert a numerical</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   variable into a factor?</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>The following aesthetics were dropped during statistical transformation: <span style="color: #00BB00;">shape</span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> This can happen when ggplot fails to infer the correct grouping structure in</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   the data.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Did you forget to specify a `group` aesthetic or to convert a numerical</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   variable into a factor?</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Square root transformation</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Wisconsin double standardization</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 0 stress 0.2133623 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 1 stress 0.2261313 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 2 stress 0.2188186 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 3 stress 0.216867 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 4 stress 0.2217179 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 5 stress 0.2159772 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 6 stress 0.2264515 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 7 stress 0.2184361 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 8 stress 0.2227609 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 9 stress 0.2298418 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 10 stress 0.2135553 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ... Procrustes: rmse 0.01747937  max resid 0.1571733 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 11 stress 0.225492 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 12 stress 0.2189902 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 13 stress 0.2164226 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 14 stress 0.2157921 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 15 stress 0.2233034 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 16 stress 0.2218725 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 17 stress 0.2201873 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 18 stress 0.2204809 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 19 stress 0.2255861 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Run 20 stress 0.2151504 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> *** Best solution was not repeated -- monoMDS stopping criteria:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      2: no. of iterations &gt;= maxit</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     15: stress ratio &gt; sratmax</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      3: scale factor of the gradient &lt; sfgrmin</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>The following aesthetics were dropped during statistical transformation: <span style="color: #00BB00;">shape</span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> This can happen when ggplot fails to infer the correct grouping structure in</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   the data.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Did you forget to specify a `group` aesthetic or to convert a numerical</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   variable into a factor?</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>The following aesthetics were dropped during statistical transformation: <span style="color: #00BB00;">shape</span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> This can happen when ggplot fails to infer the correct grouping structure in</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   the data.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Did you forget to specify a `group` aesthetic or to convert a numerical</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   variable into a factor?</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>The following aesthetics were dropped during statistical transformation: <span style="color: #00BB00;">shape</span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> This can happen when ggplot fails to infer the correct grouping structure in</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   the data.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Did you forget to specify a `group` aesthetic or to convert a numerical</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   variable into a factor?</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Done &gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt; 5.Done ordination plots </span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in adonis2(eval(parse(text = model_formula)), data = sampledf, strata = Permanova.strata,     permutations = permanova.permutation_number):</span> object 'Permanova.strata' not found</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by The package maintainer.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

